---
title: "Working with UNIX Systems"
author: "Carolina Lobato"
date: "10/6/2022"
output: html_document
---

# Statistical Genomics: Exercise 1

## 1. Navigating the keyboard

There are some keys that are used very often in Unix commands but can be
difficult to find on some keyboards. Open a text editor (eg. gedit,
nano, Vi/Vim - if you want to learn try calling vimtutor in your
terminal) and type the following keys:

**\~** tilde

**/** forward slash  

**\\** back slash or escape

**\|** vertical bar or pipe

**\#** hash, number or gate sign

**\$** dollar sign

**\*** asterisk

**'** single quote

**"** double quote

[Useful bash terminal shortcuts:]{.ul} With a little practice, you can
become very efficient navigating your computer in a CLI. Keep these
following shortcuts in mind when:

• **tab** can be used to auto fill commands or file names.

• **touch -he -- hit the enter key** will display the help page of the
command touch

• **up/down arrow** to navigate through used commands

• **history** to list last used commands

• **Ctrl+A** to move to the start of the command line

• **Ctrl+E** to move to the end of the command line

## 2. Navigating the file system

Launch a terminal window and implement the following using basic bash
commands:

```{bash}
pwd

# 1. Create a new directory in your home directory called “statistical-genomics”.
mkdir statistical-genmics

# 2. Navigate to that directory. 
cd statistical-genomics

# 3. Create a new directory in “statistical-genomics” called “Lab1”.
mkdir Lab1

# 4. What is the absolute path to “Lab1”?
pwd
  
# 5. What is the relative path from “Lab1” to “statistical-genomics”?
cd ..
  
# 6. Go back to your home directory and the root.
cd ~
cd /

# 7. Navigate from your home directory to “Lab1” using only one command.
cd ~/Projects/statistical-genomics/Lab1

# 8. Check the current directory (print its absolute path).
pwd

# 9. List all files in the current directory in human readable format (print sizes like 1K 234M 2G etc.).
ls -lha
ls -lhS
ls -lhs
ls -lht
```

## 3. Managing your files

Implement the following using basic bash commands:

```{bash}
# 1. Use your favorite text editor to create a file called “file1.txt” in your home directory.
cd ~
ls
touch file1.txt
ls

# 2. Create a directory called “test” in “statistical-genomics”.
cd Projects/statistical-genomics/Lab1
ls
mkdir test
ls

# 3. Copy “file1.txt” to the “test” directory. Name this copy “file2.txt”.
cd ~
cp file1.txt Projects/statistical-genomics/Lab1/test
mv file1.txt file2.txt

# 4. Rename “file1.txt” to “myfile1.txt”.
cd ~
mv file1.txt myfile1.txt

# 5. Move “myfile1.txt” to “test”.
mv myfile1.txt Projects/statistical-genomics/Lab1/test

# 6. Go to “test” and delete “file2.txt”.
rm file2.txt

# 7. Go to the “StatGen” directory and delete “test”. Note that there is no undo or trash folder in the terminal, so be very careful when deleting files or directories!
cd ..
rm -rf test
```

## 4. Working with text files

In this assignment, we will work with a GTF file from the Ensembl genome
browser that [describes all protein-coding and non-coding genes that
have been annotated in the human genome.]{.ul}

GTF/GFF formats are [9-column text formats used to describe and
represent genomic features]{.ul} such as exons, introns, and UTRs. The
formats have quite evolved since 1997, and despite well-defined
specifications existing nowadays, they have great flexibility and
contain a wide variety of information.

Note that the first few lines in the file beginning with a hash-tag are
so-called header lines (as opposed to data lines), and describe the
creation date, genome assembly version, etc. Header lines are not data
records.

```{bash}
pwd
# 1. curl transfers data from or to a remote file (i.e., on an FTP or HTTP site). In other words, it is the command line way of going to a website and downloading something. Download the GTF file using the “curl” command. Note that we use the “>” to redirect the data retrieved by “curl” into a new file called “human.genes.gtf.gz”:
curl http://ftp.ensembl.org/pub/release-104/gtf/homo_sapiens/Homo_sapiens.GRCh38.104.gtf.gz > human.genes.gtf.gz
  
# 2. The filename ends in “.gz”. This means that it has been compressed with a program called “gzip”. The file can be decompressed into a plain text file using the “gunzip” command:
gunzip human.genes.gtf.gz
```

After decompressing the file, the file will be called "human.genes.gtf"
-- gunzip automatically removes the ".gz" extension once it finishes
decompressing the file.

```{bash}
# Use the “head” command to print the first 5 lines of each file as output.
head human.genes.gtf
```

```{bash}
# Use the “cat” command to print all contents of the GTF file.
cat human.genes.gtf
```

```{bash}
# Use the “less” command to inspect the GTF file. Can you interrupt the process?
less human.genes.gtf
```

## 5. The "grep" command

The "grep" (global regular expression print) command is a small family
of tools that search text files for a pattern and print any lines that
match it. The basic syntax of grep is:

> grep \< OPTIONS \> PATTERNS \< FILENAME \>

```{bash}
# 1. Print all lines that start with “19” in the file “human.genes.gtf”.
grep ^19 human.genes.gtf > start19.gtf
```

```{bash}
# 2. Print three lines that come after the pattern “\sgene\s” in the file “human.genes.gtf”.
grep -A 3 "\sgene\s" human.genes.gtf > pattern3.gtf
```

```{bash}
# 3. Print all lines that do not contain the string “protein_coding” in the file “human.genes.gtf”.
grep -v protein_coding human.genes.gtf > non-prot-coding.gtf
```

## 6. The "cut" command

The "cut" command allows you to extract a specific column from a file.
By default, the column delimiter is TAB. You can change this using "-d".

```{bash}
# 1. Print the 5th byte of “human.genes.gtf ”.
cut -b 5 human.genes.gtf > byte5.gtf
```

```{bash}
# 2. Print the 2nd column of “human.genes.gtf ”.
cut -f 2 -d ' ' human.genes.gtf > column2.gtf

awk '{print $2}' human.genes.gtf | head
```

## 7. The "wc" command

The "wc" counts the number of lines/bytes/characters/words in a file.

```{bash}
# 1. How many words does “human.genes.gtf ” contain?
wc -w human.genes.gtf
```

```{bash}
# 2. How many lines does “human.genes.gtf ” contain?
wc -l human.genes.gtf
```

## 8. Piping

A pipe ( \| ) is a form of redirection (transfer of standard output to
some other destination) that is used to send the output of one command
to another command for further processing.

• You can redirect output to a file using **\>**

• To append use \>\>

Use a single command line for each of the following:

```{bash}
# 1. To print the headers of “human.genes.gtf ” we can pipe the output of cat and use grep to match a hashtag
cat human.genes.gtf | grep \#
```

```{bash}
# Can you think of a quicker way to get this result?
head human.genes.gtf | grep \# > head.gtf
```

```{bash}
# 2. How many GTF records (i.e., data lines) are in “human.genes.gtf ”?
grep -v \# human.genes.gtf | wc -l 
```

```{bash}
# 3. How many GTF records (i.e., data lines) in “human.genes.gtf ” correspond to protein-coding genes?
grep -v \# human.genes.gtf | grep "\sgene\s" | grep -c protein_coding

grep "gene_biotype "protein_coding"
```

```{bash}
# 4. Create a new file temp1.txt containing the last 5 lines of “human.genes.gtf ”, but only showing the 1st and 3rd columns.
tail -n 5 human.genes.gtf | cut -f 1,3 > temp1.txt
```

```{bash}
# 5. Create a new file temp2.txt containing the first 3 lines of 2nd column of temp1.txt
head -n 3 temp1.txt | cut -f 2 > temp2.txt
```

```{bash}
# 6. Add the first 10 lines of “human.genes.gtf ” if they contain the string “protein_coding”
head -n 10 human.genes.gtf | grep protein_coding >> temp2.txt
```

```{bash}
# 7. How many GTF records (i.e., data lines) in “human.genes.gtf” correspond to exons from protein-coding genes?
grep protein_coding human.genes.gtf | grep "\sexon\s" | grep -c protein_coding
```

```{bash}
# 8. How many GTF exons from protein-coding genes are on the forward (+) strand?
grep protein_coding human.genes.gtf | grep "\sexon\s" | grep -c +
```

```{bash}
# 9. How many GTF exons from protein-coding genes are on the reverse (-) strand?
grep protein_coding human.genes.gtf | grep "\sexon\s" | cut -f 7 | grep -c "-"
```

```{bash}
# 10. How many CoDing Sequence (CDS) exons (records where column 3 is “CDS”) from protein-coding genes exist on each chromosome (column 1)?
grep protein_coding human.genes.gtf | cut -f 1,3 | grep "CDS" | uniq -c | head -23

$i # for loops
for C in {1..22}
echo "Chromosome$C"
done
```

```{bash}
# 11. Print recent commands you typed for this exercise that contained the command cat
history | grep cat
```

Explain how you might design an analysis of "human.genes.gtf " that
would reveal [how many distinct protein-coding genes]{.ul} there are in
the human genome.

Hint: you may not have learned all of the command you might need -- the
point is to think about what what you could do with the commands you
know of and what limitations would have to be addressed.

```{bash}
grep protein_coding human.genes.gtf | cut -f 3 | uniq -c

# uniq applies to non-conseutive interspersed lines
```

## 9. Writing a bash script

When you start dealing with multiple files, manipulating them one by one
will become tedious. We can automate the execution of our commands by
saving them in a file **\<file.sh\>**.

Each permission is represented by a single letter- r (read), w (write),
and x (execute) for each ownership level (user, group, others). Use
chmod +x \<filename.sh\> to add executable permission and recheck file
permissions.

```{bash}

# 1. Using any text editor create a new file named “hello-world.sh”. The first line will contain a shebang "#!" which defines our script’s interpreter as bash.
#!/bin/bash
echo "Hello World"

# 2. Save the file and make sure your script is executable. This is indicated by the first column in the ls -l <filename.sh>.

# 3. Execute the bash script either with bash hello-world.sh or ./hello-world.sh

```

If you want to learn more about UNIX, this intro course is very helpful:
<http://korflab.ucdavis.edu/unix_and_perl/>
