---
title: "RNA-seq Data Analysis"
author: "Carolina Lobato"
date: "11/16/2022"
Outputs: html_document
---

# Statistical Genomics: Exercise 5

## 3. Mapping RNA-seq reads with HISAT2

##### 1. To efficiently map the reads, the reference genome must first be indexed. Indexing takes relatively long, but only needs to be performed once. You will be working with chromosome. Enter the following command in the Linux terminal:

```{bash}
cd Lab5/Tools/hisat2-2.2.1
chmod +x hisat2-build
```

```{bash}
cd Lab5/Data
gunzip Arabidopsis_thaliana.TAIR10.dna.chromosome.1.fa.gz
```

```{bash}
mkdir -p Outputs/hisat2
Tools/hisat2-2.2.1/hisat2-build Data/Arabidopsis_thaliana.TAIR10.dna.chromosome.1.fa Outputs/hisat2/Arabidopsis_thaliana.TAIR10.dna.chromosome.1.fa
```

What are the names and paths of the Outputs files generated by hisat2-build?

```{bash}
ls Outputs/hisat2/Arabidopsis_thaliana.TAIR10.dna.chromosome.1.fa.*
```

##### 2. Once the genome has been indexed, you can map the reads in each FASTQ file with HISAT2. HISAT2 is a program that aligns RNA-seq reads rapidly to a genome. Run HISAT2 with the following parameters:

• -q : Defines the input file as fastq files.

• -x : The indexed genome prefix.

• -U : The single-end input read file in fastq format.

• -S : The Outputs file with .sam extension.

Create a folder called "results" in the same folder in which you have stored the Data. This is where you should store the mappings.

The following bash script will run HISAT2 on all the .fastq files in the working directory:

```{bash}
mkdir Outputs/sam
files='SRR064154.fastq SRR064155.fastq SRR064166.fastq SRR064167.fastq'
for input in $files
do
Tools/hisat2-2.2.1/hisat2 -q -x Outputs/hisat2/Arabidopsis_thaliana.TAIR10.dna.chromosome.1.fa -U Data/$input -S Outputs/sam/$input".sam"
done
```

What are the names and paths of the Outputs files generated by HISAT2?

```{bash}
ls Outputs/sam/*.sam
```

For follow-up steps, Outputs .sam files need to be transformed into .bam files using samTools. You can use the following code for that:

```{bash}
mkdir -p Outputs/bam
files='SRR064154.fastq SRR064155.fastq SRR064166.fastq SRR064167.fastq'
for input in $files
do
Tools/samTools-1.16.1/samTools view -bS Outputs/sam/$input".sam" > Outputs/bam/$input".bam"
done
```

## 4. Counting the reads overlapping with certain genomic features

Given a file with mapped sequencing reads and a list of genomic features, a common task is to count how many reads map to each feature. In the case of RNA-Seq, the features are typically genes or the union of their exons. One may also consider each exon as a feature and quantify, for example, alternative splicing.

### 4.1. Importing annotation Data into R

The *rtracklayer* R package includes several useful functions to import/export *GRanges* objects from and to different Data formats commonly used in genomic analyses. *GRanges* objects define integer ranges (IRanges) with two more important pieces of information:

• the chromosome or scaffold name (seqnames)

• the DNA strand

The DNA strand can be specified as plus "+" or minus "-", or left unspecified with "\*". Thus *GRanges* are objects representing genomic interavals.

```{r}
#install.packages("BiocManager")
BiocManager::install("rtracklayer")
```

```{r}
library("rtracklayer")
gtf <- import("Lab5/Data/Arabidopsis_thaliana.TAIR10.48.gtf.gz")
class(gtf)
head(gtf, 2)
```

Select only the protein-coding gene features:

```{r}
# Exclude NA values
idx <- !is.na(mcols(gtf)$type) & !is.na(mcols(gtf)$gene_biotype)
genes_gtf <- gtf[idx]
idx <- mcols(genes_gtf)$type == "gene" & mcols(genes_gtf)$gene_biotype == "protein_coding"
genes_gtf <- genes_gtf[idx]
```

Finally, split the GRanges object by gene identifier to produce a GRangesList object. Use this object to compute the number of reads in each sequencing library overlapping each gene in the Arabidopsis thaliana genome:

```{r}
genes_gtf <- split(genes_gtf, mcols(genes_gtf)$gene_id)
genes_gtf
```

### 4.2. Counting reads with summarizeOverlaps

Read the "targets.txt" file into R:

```{r}
targets <- read.delim("Lab5/Data/targets.txt")
targets
```

To count the reads, you will use the *summarizeoverlaps()* function *GenomicAlignments* R package. The mapped reads must be passed as arguments of the *summarizeoverlaps()* function. This can be done in several ways, including as a BamFileList object.

```{r}
library("RsamTools")
samples <- as.character(targets$FASTQFile)
samplespath <- paste("./Lab5/Outputs/bam/", samples, ".bam", sep="")
names(samplespath) <- samples
bfl <- BamFileList(samplespath, yieldSize=50000, index=character())
```

The RsamTools R package is an interface between R and SamTools. Now, you can count the reads overlapping with genomic features in genes_gtf:

```{r}
library("GenomicAlignments")
countDF <- summarizeOverlaps(genes_gtf, bfl, mode="Union", ignore.strand=TRUE)
countDF
```

• "mode" can be one of the pre-defined count methods such as "Union", "IntersectionStrict", or "IntersectionNotEmpty" or it a user supplied count function.

-- "Union": (Default) Reads that overlap any portion of exactly one feature are counted. Reads that overlap multiple features are discarded.

*summarizeOverlaps()* returns a *RangedSummarizedExperiment* object. The assays slot holds the counts:

```{r}
countDF <- assays(countDF)$counts
head(countDF, n=3)
```

• Save the read counts as a table.

## 5. Testing for differential expression with the DESeq2 R package

Differential expression analysis is used to identify differences in the transcriptome [across a cohort of samples]{.underline}. There are many Tools available to perform this type of analysis. *DEseq2* is a popular differential expression analysis R package. Its differential expression tests are based on a [negative binomial generalized linear model]{.underline}.

```{r}
library("DESeq2")
```

Input Data for DEseq2 consists of (non-normalized) read counts at either the gene or transcript level.

### 5.1. Experimental design

First, you need to specify which factor(s) in the metaData are relevant for the experimental design and how they should be used in the analysis. The formula it should take the form of a " ∼ " followed by "+" separating factors.

Let us assume that you are interested in evaluating differences between the translatome and the transcriptome: **∼ Condition**

Note that the control or reference sample group for a given factor should appear first when viewing the levels of the corresponding variable. Here, you have only one factor of interest ("Condition") and two sample groups ("AP3" and "TRL"), and you will be **using "AP3" as control**:

```{r}
targets$Condition <- as.factor(targets$Condition)
levels(targets$Condition)
```

Now you have everything you need to perform a differential expression analysis, namely:

1\. a table with the counts for each gene and each RNA-seq library;

2\. a table with metaData describing the samples;

3\. an experimental design formula.

The next step is to create an *DESeqDataSet* object, which will **store the read counts and intermediate calculations needed for the differential expression analysis**. The object will also **store the design formula** **used to estimate the dispersion and the fold-changes used to identify differentially expressed genes**.

```{r}
dds <- DESeqDataSetFromMatrix(countData=countDF, colData=targets, design=~Condition)
dds
```

Note that DESeq2 assumes that the samples in the count table (columns) are in the same order as in the table with the descriptions (rows).

### 5.2. Running the DESeq2 pipeline

The next step is to run the *DEseq()* function on our DESeq2 Data set object. In this step the algorithm will perform the following:

1\. estimate the size factors;

2\. estimate the dispersion;

3\. fit a Negative Binomial generalized linear model for each gene and compute the Wald statistic.

```{r}
dds <- DESeq(dds)
```

Indeed, DEseq() is a wrapper for three functions, which can also be executed separately:

```{r}
# dds <- estimateSizeFactors(dds)
# dds <- estimateDispersions(dds)
# dds <- nbinomWaldTest(dds)
```

DESeq2 computes a P-value for each gene:

• The null hypothesis H0 states that "there is no effect of the treatment on the gene and that the observed difference between treatment and control is merely caused by experimental variability".

• The alternative hypothesis H1 is that the gene is differentially expressed between the treatment and the control.

### 5.3. Inspecting the results

The results for the last variable in the design formula can be extracted using the *results()* function. In our case, there is only one, **Condition**:

```{r}
res <- results(dds) # res is a Data.frame with several columns
mcols(res)
res
```

```{r}
alpha <- 0.05 # Threshold on the adjusted p-value
cols <- densCols(res$log2FoldChange, -log10(res$pvalue)) # densCols produces a vector containing colors which encode the local densities at each point in a scatterplot.
plot(res$log2FoldChange, 
     -log10(res$padj), 
     col=cols, 
     panel.first=grid(),
     main="Volcano plot", 
     xlab="Effect size: log2(fold-change)", 
     ylab="-log10(adjusted p-value)",
     pch=20, 
     cex=0.6
     )
abline(v=0)
abline(v=c(-1,1), col="brown")
abline(h=-log10(alpha), col="brown")

gn.selected <- abs(res$log2FoldChange) > 2.5 & res$padj < alpha 
text(res$log2FoldChange[gn.selected],
     -log10(res$padj)[gn.selected],
     lab=rownames(res)[gn.selected ],
     cex=0.4
     )
```
