---
title: "Gene Expression Analysis with Microarrays"
author: "Carolina Lobato"
date: "10/23/2022"
output: html_document
---

```{r loadlib, echo = FALSE}
library (dplyr)
library (tidyr)
library (ggplot2)
```

# Statistical Genomics: Exercise 3

## Affymetrix GeneChips

### 2. Importing Data

The intensities measured for each microarray are normally stored in a .CEL file. This file contains an intensity value for each physical location (cel) of the microarray. The information needed to map each physical location to a probe set is described in a .CDF file.

#### 2.1. The Gene Expression Omnibus

GEO: www.ncbi.nlm.nih.gov/geo/

This is a public repository that archives and freely distributes microarray, next-generation sequencing, and other forms of high-throughput genomic data submitted by the scientific community.

We will work a with a microarray dataset of HUVEC and ocular vascular endothelial cells generated by Andrew C. Browning et al.: <https://pubmed.ncbi.nlm.nih.gov/22028475/>

###### 1. What was the aim of the study by Browning et al.?

> To investigate the difference between human umbilical vein endothelial cells (HUVEC) and human ocular microvascular endothelial cell (MVEC) gene expression, and to determine if these differences could improve the understanding of ocular angiogenic diseases.

Open your favorite Web Browser, go to <https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE20986> , and download the .CEL files for the GEO accession numbers in the table. Note that the entire dataset consists of 12 microarrays, but we will only be using 6.

###### 2. Create and go into a folder " GSE20986 ".

```{bash}
pwd
mkdir GSE20986
ls
cd GSE20986
pwd
```

###### 3. Download the data set either by going to the appropriate GEO web-page or use following script to download via ftp.

```{bash}
#!/bin/bash

SERIES = "/Lab3/Data/GSE20986"
wget https://ftp.ncbi.nlm.nih.gov/geo/series/GSE20nnn/"$SERIES"/suppl/"$SERIES"_RAW.tar

tar -xf "$SERIES"_RAW.tar

rm $SERIES_RAW.tar

META_DATA = "/Lab3/Data/phenodata.txt"

cat "$META_DATA"

while read line;do
  if [[ $line =~ ^GSM ]]; then
    SAMPLE = $( echo $line | sed "s/CEL.*/CEL/")
    echo $SAMPLE
    gunzip $SAMPLE.gz
  fi
done < $META_DATA

rm * gz
```

###### 4. If you have downloaded the files manually, you will need to generate the meta-file, describing the experimental factors yourself using your favorite plain text editor. Use tab character to separate the fields. Save it as a plain text file called "phenodata.txt".

###### 5. The .CEL files can be imported into R using the "simpleaffy" R/Bioconductor package:

[https://www.bioconductor.org/packages//2.7/bioc/html/simpleaffy.html](https://www.bioconductor.org/)

This package also provides access to a variety of metrics for assessing the quality of RNA samples and of the intermediate stages of sample preparation and hybridization.

hgu133a2.db contains annotation data of the chip. [ALSO NEEDS INSTALLATION]

The compatible R version has to be installed using conda in a different environment than the operating system. [BUT IT DOES NOT WORK: USING "affy" PACKAGE INSTEAD]

The link in TC wrong, go to:

<https://astrobiomike.github.io/R/managing-r-and-rstudio-with-conda>

```{bash}
conda search -c conda-forge r-base
conda create -n R-4.0.5 -y -c conda-forge r-base=4.0.5 rstudio r-tidyverse
conda activate R-4.0.5
```

```{r, installation}
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install(version = "3.12")
BiocManager::install("simpleaffy")
```

###### 6. Use the function read.affy() to read the .CEL and " phenodata.txt " files into an object named "celfiles". "celfiles" is an " AffyBatch " object. You can verify it using the class() function.

```{r}
# browseVignettes("simpleaffy")
library(simpleaffy)
getwd()
celfiles <- read.affy("/Lab3/Data/phenodata.txt", "/Lab3/Data/GSE20986")
class(celfiles)
```

###### 7. Obtain a summary of the celfiles object by typing its name.

```{r}
celfiles
```

• Based on the information obtained, can you find the product page on the website of the microarray manufacturer? What organism is this product for? Which genes are represented on the microarray?

> <https://www.thermofisher.com/document-connect/document-connect.html?url=https://assets.thermofisher.com/TFS-Assets%2FLSG%2Fbrochures%2Fhgu133arrays_datasheet.pdf>
>
> GeneChip® **Human Genome** U133 Plus 2.0 Array (single array, cartridge format) - Analyze **gene expression across the entire human genome** for one sample.
>
> Analyzes the relative expression level of more than 47,000 transcripts and variants, including more than 38,500 well characterized genes and UniGenes.
>
> [\<https://www.affymetrix.com/analysis/netaffx_analysis_center_retired.html\>?](https://www.affymetrix.com/analysis/netaffx_analysis_center_retired.html?) The information provided by the NetAffx Analysis Center is available in the annotation files and sequence files accessible from the [Thermo Fisher microarray product pages](https://www.thermofisher.com/us/en/home/life-science/microarray-analysis.html). [IT IS NOT]

• How many samples have you read?

> number of samples = 6

• How many genes are represented on each microarray? Are these really genes?

> number of genes = 54675
>
> Reverse transcribing the mRNA into complementary DNA (cDNA).
>
> Could be the gene itself, or a segment of it.

In addition to the intensities, the AffyBatch object contains information about the microarray experiments.

• What are these three functions doing? What kind of information do they provide?

```{r}
phenoData(celfiles) 
# returns an object containing information on both variable values and variable meta-data.
pData(celfiles)
# pData returns a data frame with samples as rows, variables as columns. 
varLabels(celfiles) 
# returns a character vector of measured variables. 
```

### 3. Quality Control

As with other high-throughput technologies, quality control (QC) is a fundamental aspect of the analysis process. Data quality can be checked using various diagnostic plots.

#### 3.1. Visual Inspection

The image() function creates an image of colors or shades of gray that represent the intensities for each spot on a microarray. This image can be used to explore any spatial effects across the microarray. For some examples of images showing different artifacts, have a look at: <http://plmimagegallery.bmbolstad.com/>.

###### 8. Examine all the microarrays. Do you see any obvious problems?

> Spatial correction: The signal is expected to be similarly distributed across the entire microarray. Spatial effects can be removed by smoothing the intensity data across the microarray. Then, a smooth surface can be fit to the data and subsequently subtracted from the data.

```{r}
image(celfiles)
# There is a systematic scratch in all images
# GSM524665.CEL has missing reads in the bottom
# GSM524667.CEL has a different resolution 
```

#### 3.2. Probes & Probe sets

##### 3.2.1. Probes

In principle, the MM probes provide a measure of cross-hybridization to the corresponding PM probes. Cross-hybridization is the tendency for a nucleic acid sequence to bind to other nucleic acid sequences that are almost, not perfectly complementary. [Hence, subtracting the signal intensities of MM probes from those of PM probes should allow canceling of the effect of cross-hybridization]{.underline}. The intensities of the PM (MM) probes can be accessed with the **pm()** (**mm()**) functions.

###### 9. How many PM probes are present on each microarray?

###### 10. How many MM probes do you expect on each microarray? Now check how many MM probes are there on each microarray.

```{r}
nrow(pm(celfiles)) # 604258
nrow(mm(celfiles)) # 604258
```

> PM and MM are expected to be the same number.

Typically, probe intensities are log-transformed. This is done for several reasons: [it stabilizes the variance; it compresses the range of the data; and it makes the data more normally distributed.]{.underline}

###### 11. Generate histograms of the intensities of the PM and MM probes for each microarray. Do you see any difference in the distributions of the PM and MM probes? Is this expected? Why?

> MM has higher counts in lower intensities. This is expected as cross hybridization should not be higher or as high as PM.

```{r}
as.data.frame(pm(celfiles)) %>% 
  pivot_longer(cols = 1:6, names_to = "Microarray", values_to = "Intensity") %>%
  ggplot() + geom_histogram(aes(x = Intensity)) + scale_x_log10() + facet_grid(cols = vars(Microarray)) + labs(y = "Counts",
title ="PM probe itensity")

```

```{r}
as.data.frame(mm(celfiles)) %>% 
  pivot_longer(cols = 1:6, names_to = "Microarray", values_to = "Intensity") %>%
  ggplot() + geom_histogram(aes(x = Intensity)) + scale_x_log10() + facet_grid(cols = vars(Microarray)) + labs(y = "Counts",
title ="MM probe itensity")
```

##### 3.3.2. Probe sets

You can access the probe set identifiers through the geneNames() function. Note that these are not gene names!

###### 12. How many probe sets are on each microarray?

```{r}
summary(geneNames(celfiles)) #  54675 probe sets
```

Given one or several probe set identifiers, you can extract the corresponding intensities using the probeset() function.

###### 13. How many probe pairs are there for "1552277_a\_at"? The probe set "1552277_a\_at" represents the MSANTD3 gene (<https://www.genecards.org/cgi-bin/carddisp.pl?gene=MSANTD3>). Little is known about this gene, except that it is associated with swine influenza infections.

```{r}
probeset(celfiles, genenames = "1552277_a_at") # 11 probes
```

Probe set are S4 objects consisting of three slots:

• "id": the probe set identifier.

• "pm": a matrix with the PM probe intensities.

-- Columns represent samples; rows represent probes.

• "mm": a matrix with the MM probe intensities.

-- Columns represent samples; rows represent probes.

###### 14. How many probe pairs are there for "1007_s\_at"? Extract the PM and MM intensities for "1007_s\_at".

```{r}
pmat <- probeset(celfiles, genenames = "1007_s_at")[[1]]@pm
nrow(pmat) # 16

mmat <- probeset(celfiles, genenames = "1007_s_at")[[1]]@mm
nrow(mmat) # 16
```

###### 15. Use the barplot.ProbeSet() function to visualize the intensities for the PM and MM probes of probe set "1552277_a\_at".

```{r}
probeset(celfiles, genenames = "1552277_a_at")[[1]] %>% barplot.ProbeSet(ask = FALSE, col.pm = "red", col.mm = "blue") 
```

###### 16. How are the intensities for the PM probes compared to those of the MM probes? What does this suggest for the probe set (and the corresponding gene)?

> PM intensities are higher for all probes in all microarrays indicating expression and low cross-hybridization. There is also consistency between probe pairs within the microarray.

• Examine the intensities of the probes for probe set "1007_s\_at". What can we conclude about the expression of this probe set (and the corresponding gene)?

```{r}
probeset(celfiles, genenames = "1007_s_at")[[1]] %>% barplot.ProbeSet(ask = FALSE, col.pm = "red", col.mm = "blue") 
```

> PM and MM intensities are similar for most probes in all microarrays indicating high cross-hybridization and low expression of the gene in all samples. Consistency between probe pairs within the same microarray is also not very high.

#### 3.3. Comparison Between Microarrays

##### 3.3.1. Average Background

The average background of a given microarray is computed as the [2nd percentile of the PM and MM probe intensities]{.underline} in a given region of the microarray. This summary statistic is intended to measure optical background. Average background values ranging between 20 and 100 are typical for a good quality microarray.

###### 17. What could lead to a higher average background?

> The average background may depend upon biological variation, experimental or technical variation (array-specific effects, gene-specific effects, dye-specific effects, background noise and artifacts, and preparation effects).

##### 3.3.2. Scale factors

The Bioconductor packages usually contain shortcuts to visualize complicated data types using common functions. For example, we can use the **hist()** function to visualize the distribution of the data:

```{r}
hist(celfiles, which = "both", col = 1:6, lty = 1, lwd = 1)
```

###### 18. Similarly, you can create a boxplot using the **boxplot()** function. Try it.

```{r}
boxplot(celfiles)
```

###### 19. Are the intensities on the different microarrays comparable?

> All microarrays seem to be comparable (except GSM52665.CEL).

Assuming that all other factors remain constant, the overall intensity observed for a given microarray reflects how much labeled target was hybridized to the probes. [The median intensity of all PM and MM probes on a microarray are used to compute a summary statistic known as scale factor.]{.underline} Since the scale factor provide a measure of the overall intensities observed for a microarray, [large variations in the scale factors of analyzed microarrays might indicate issues with RNA extraction, labeling, scanning or even microarray manufacture.]{.underline} Within an experiment, arrays are expected to have scale factors within **3-fold of each other**; arrays whose scale factors are outside this range are considered to have poor quality.

##### 3.3.3. MA plots

A MA plot is a scatter plot designed to detect differences between the labels in two-color microarrays. In two-color microarrays, two biological samples are labeled with different fluorescent dyes, usually Cyanine 3 (Cy3) and Cyanine 5 (Cy5), and then hybridized to the same array. M and A are defined as follows:

M = log 2 (R/G) = log 2 (R) − log 2 (G)

A = 1/2 log 2 (RG) = 1/2 (log 2 (R) + log 2 (G))

[R is the Cy5 intensity]{.underline} and [G is the Cy3 intensity.]{.underline}

**M is plotted against A**.

• Affymetrix GeneChips use single-color labeling.

Given a microarray, **M** is the difference between the (logarithm of the) intensity of a probe on the microarray and the median (logarithm of the) intensity of the probe across all analyzed microarrays. **A** is the average of the (logarithm of the) intensity of a probe on the microarray and the median (logarithm of the) intensity of that probe across all analyzed microarrays. A **MA plot** identifies intensity-related biases.

###### 20. Produce a MA plot for each microarray using the MAplot() function.

```{r}
getwd()
celfiles2 <- read.affy("/Lab3/Data/phenodata2.txt", "/Lab3/Data/GSE20986")
MAplot(celfiles2)
```

• The red line is a locally estimated scatterplot smoothing (loess) regression curve based on the data. Ideally, the cloud of data points in the MA-plot should be centered around M=0 (blue line). What is the assumption?

> The assumption is that a probe's intensity won't systematically deviate from the median intensity verified across all analysed microarrays.

#### 3.4. Hybridization

##### 3.4.1. 3' to 5' ratios

Affymetrix's standard labeling protocols rely on a reverse-transcription step. Reverse transcription proceeds from the 3' end of a sequence. Therefore, RNA degradation or problems during labeling, often lead to under representation at the 5' end of the sequence.

The mRNA transcripts of **GAPDH** (<https://www.genecards.org/cgi-bin/carddisp.pl?gene=GAPDH>) and **ACTB** (<https://www.genecards.org/cgi-bin/carddisp.pl?gene=ACTB>) are relatively long and abundant in most cell types. Most Affymetrix microarrays contain separate probe sets targeting the **5', mid and 3' regions** of these transcripts. By comparing the intensity values of the 3' probe set to either the mid or 5' probe set, it is possible to obtain a measure of the quality of the RNA hybridized to the microarray.

• High ratios indicate RNA degradation or problems during labeling. GAPDH is shorter than ACTB.

-- The ratio for GAPDH should be at or around 1 (and always less than 1.25).

-- The ratio for ACTB should be below 3 to be considered acceptable.

• All analyzed microarrays should have ratios in a similar range.

Assess the average background, scale factors, and 3' to 5' ratios of the microarrays using the **qc()** function. The results can be visualized using the **plot()** function.

```{r}
quality <- qc(celfiles)
plot(quality)
```

In the resulting figure:

• The fraction of probe sets that are called present on each microarray (top) and the average background intensities (bottom) are written along the left hand side of the figure. In order to determine whether a given probe set is present, its PM and MM probe intensities are compared using one-sided Wilcoxon's signed rank test. If any microarray has a substantially different value compared to the others, the numbers will be shown in red, otherwise they will be blue.

-- GAPDH ratios are plotted as circles. Values below 1 are displayed in blue; values above 1, in red.

-- ACTB ratios are plotted as triangles. Values below three (recommended) are given in blue, those above three in red.

• The central blue stripe represents the range where scale factors are within 3-fold of the mean across the scale factors of all analyzed microarrays: The scale factors are plotted as horizontal lines departing from the central vertical line of the figure. A line to the left corresponds to a down-scaling, to the right, to an up-scaling. If any scale factors fall outside this 3-fold region, they are all colored red, otherwise they are blue.

###### 21. Why should the fraction of probe sets that are called present on each microarray be similar among the different microarrays? How do you interpret your results?

> Because these are abundant endogenous probe sets that should be expressed in all microarrays and in all samples. According to the plot both GAPDH and ACTB rations are within the acceptable parameters in all microarrays, including sample GSM524663.

##### 3.4.2. Spike-in probe sets

In addition to the probe sets for the endogenous GAPDH and ACTB, Affymetrix GeneChips contain probe sets for **BioB**, **BioC**, **BioD** and **CreX**. BioB, BioC, BioD and CreX are Bacillus subtiliis genes and no human mRNA transcript is supposed to bind them. The mRNA transcripts of BioB, BioC, BioD and CreX are added (spiked in) into the samples during their preparation.

• BioB is added at a concentration of 1.5 pM, corresponding to approximately three transcripts per cell, the lower limit of detection for the system.

• BioC, BioD and CreX are spiked-in at increasing concentrations.

###### 22. Use the spikeInProbes() function to summarize the intensities of the spike-in probe sets, and the matplot() function to visualize them. How do you interpret your results?

```{r}
spikeInProbes(quality) %>% matplot()
```

> Intensities of each spike in probe set is homogenious between microarrays. Exceptions can be probe set 3 in microarray 4 and probeset 1 in microarray 5. Assuming probe set 1 to be BioB this may indicate a higher sensitivity of microarray 5. Microarray 4 seems to have a lower detection of at the level of concentration of BioD.

#### 3.5. Expression Variability

##### 3.5.1. RLE Plots

For a given microarray, the RLE (Relative Log Expression) value of a probe set is the [ratio between the expression value of the probe set and the median expression value of the probe set across all microarrays analyzed]{.underline}. The ratios are then log-transformed. The expression value of a probe set is estimated using robust regression, as a [function of probe and microarray effects]{.underline}. Basically, **the aim is to down-weight outlier probes**. This is implemented in the "affyPLM" (<https://www.bioconductor.org/packages/release/bioc/html/affyPLM.>html) R/Bioconductor package.

###### 23. Use the "fitPLM()" function to estimate the probe set expression values.

```{r}
# BiocManager::install("affyPLM", force=TRUE, configure.args="--disable-threading")
# BiocManager::install("preprocessCore", force = T, configure.args="--disable-threading")
library(affyPLM)
rle.values <- fitPLM(celfiles)
```

###### 24. Visualize the distribution of the RLE values of the probe sets of each microarray using the Mbox() function.

```{r}
Mbox(rle.values)
```

###### 25. Assuming that the expression levels of most genes (and, hence, their probe sets) will remain constant during an experiment, what should be the median of each distribution in an ideal experiment? What about the spread of the distribution? How should it look like in an ideal experiment? Do any of your microarrays deviate from the expectation?

> The median of each distribution in an ideal experiment should be 0 and have a normal distribution. Only GSM524665.CEL slightly deviates.

##### 3.5.2. NUSE plot

The Normalized Unscaled Standard Error (NUSE) plot [visualizes the distribution of the standard errors of the model]{.underline}. On the NUSE plot, the standard errors are standardized so that for each probe set, the median standard error across all analyzed microarrays is equal to 1. As a rule of thumb, a **median equal to or greater than 1.05 indicates quality problems**, which are often also detected by other QC analyses.

###### 26. Produce a NUSE plot and interpret your results. Do any of your microarrays deviate from the expectation?

```{r}
NUSE(rle.values)
```

> Any of the microarrays has a medianequal to or greater than 1.05.

### 4. From Intensities to Expression Values

Many background correction and normalization algorithms (MAS5, gcRMA, RMA) have been developed to correct for non-biological differences in microarray signals. These algorithms employ different methods for background correction, normalization, and probe set summarization.

#### 4.1. Robust Multi-array Average (RMA)

The Robust Multi-array Average (RMA) algorithm summarizes each probe set as a robust multi-array average of log-transformed, background-adjusted, normalized PM intensities. **The resulting value is known as the expression value of the probe set.** RMA ignores MM probes entirely: although the MM probes provide information, implementing this without adding noise is challenging.

**(i)** **Background correction:** Let *F~ijk~* be the foreground intensity for the *j^th^* PM probe in the *k^th^* probe set on the *i^th^* microarray and *B~ijk~* be the corresponding background intensity, then the [background-adjusted intensity]{.underline} of the PM probe is

*PM~ijk~ = F~ijk~ + B~ijk~* where *F~ijk~ ∼ e^λ~ijk~^* and *B~ijk~ ∼ N (μ~i~ , σ^~2~^~ijk~)*

The estimates for *μ~ijk~* , *σ~ijk~* and *λ~ijk~* are computed for each microarray using the distribution of PM probe intensities by fitting the model

*E[F~ijk~ \|PM~ijk~] = PM~ijk~ − μ~i~ − λ~ijk~ σ^~2~^~ijk~* *\> 0*

**(ii)** **Normalization between arrays:** Between-array normalization is performed using quantile normalization (<https://en.wikipedia.org/>wiki/Quantile_normalization).

**(iii)** **Summarization of the probe sets:** The probe sets are summarized as the robust average of their log-transformed, background-adjusted, normalized PM probe intensities using the median polish algorithm (<https://en.wikipedia.org/wiki/>Median_polish). The median polish algorithm is a type of robust 2-way ANOVA, where one factor is the microarray and the other is the probe set. The algorithm is robust to outliers, so that single probes with large values are down-weighted. Because both the quantile normalization and the median polish algorithms use the data from all the microarrays, using just a subset of the microarrays or removing a single bad microarray affects the normalization step for all microarrays.

The resulting value is expression value of the probe set. Apply the RMA algorithm as implemented in the (**rma()**) function:

###### 27. What object is returned by the rma() function? Which slots does the object contain ? Ask for some help about this class of objects.

```{r}
eset <- rma(celfiles)
class(eset) # ExpressionSet with assayData, phenoData, featureData, experimentData, annotation, protocolData	

?Biobase()

```

The **exprs()** function returns the expression values of the probe sets as a matrix:

```{r}
rma.exprs <- exprs(eset)
```

Following background correction, normalization, and probe set summarization, diagnostic plots should be used to assess the quality of the data.

###### 28. Visualize the distribution of the expression values using the boxplot() function.

```{r}
boxplot(rma.exprs)
```

###### 29. What is the difference between these boxplots and the ones in 18?

> In this boxplot the intensities of the microarray GSM524665 was corrected and the outliers were identified.

###### 30. Do the background correction and normalization procedures correct all the differences between the analyzed microarrays?

> No, outliers are still present.

Depending on the QC assessment, you may decide to discard some of the data before proceeding with the differential expression analysis.

###### 31. Would you discard any of the microarrays based on your previous results?

> No.

### 5. Differential Expression Analysis

The goal of differential expression analysis is to identify genes whose expression differs under different conditions. We assume a linear model ***Y = Xβ + ε***

where ***Y*** **is a vector containing the expression values** of the probe sets, ***X*** **is the design matrix**, ***β*** **is a vector of regression coefficients** and ***ε*** **is a normally distributed error** term.

The choice of design matrix is a critical step in linear modeling since it [encodes which coefficients will be fit in the model]{.underline}, as well as the [inter-relationship between the samples]{.underline}.

Here we will compare between two groups of samples: **iris versus retina**. We encode this experimental design in R with two steps:

**(i)** We start a formula with the tilde symbol (**∼**).

**(ii)** Then we write the name of the variable that separates the samples into two groups.

###### 32. In our case, the variable is called "Target":

```{r}
X <- model.matrix(~ celfiles$Target)
```

This design matrix has two columns because there are two parameters in this conceptual design: [the expression level in each of the two groups]{.underline}. Column 1 represents the expression of the iris group and column 2 represents [the difference]{.underline} in expression level from the retina group to the iris group. Testing that the two groups have the same expression level is done by testing whether **the second parameter** (equal to the difference in expression between the two groups) **is equal to zero.**

A different parametrization is:

```{r}
X <- - model.matrix(~ celfiles$Target-1)
colnames(X) <- c("iris", "retina")
```

In this design matrix, the two parameters corresponding to the two columns r**epresent the expression levels in the two groups**. Testing that the two groups have the same expression level is done by [asking whether these two parameters are the same]{.underline}.

###### 33. We will use the limma R/Bioconductor (<https://www.bioconductor.org/packages/release/>bioc/html/limma.html) package to fit a linear model to each probe set.

```{r}
# BiocManager::install("limma")
library(limma)
fit1 <- lmFit(eset, X)
```

###### 34. Significance testing can be done for each parameter, to which limma refers as "contrasts". We are interested in an association between probe set expression and tissue of origin:

```{r}
contrast.matrix <- makeContrasts(irisvsretina = "iris-retina", levels = X)
```

###### 35. Given a linear model and a set of contrasts, we can now compute the estimated coefficients and standard errors for that particular set of contrasts:

```{r}
fit2 <- contrasts.fit(fit1, contrast.matrix)
```

###### 36. Finally, we compute the moderated t-statistic, empirically shrinking the residual variance of each probe set:

```{r}
fit3 <- eBayes(fit2)
```

and use the function **topTable()** to extract the results. The column "adj.P.Val" contains false discovery rate (FDR) adjusted P-values:

```{r}
topTable(fit3, coef = 1, n = 5, adjust.method = "fdr")
```

###### 37. At a significance level of 0.05, how many probe sets are differentially expressed?

```{r}
sum(fit3$p.value <= 0.05) # non-adjusted = 13034
nrow(topTable(fit3, n = length(fit3$p.value), p.value = 0.05, adjust.method = "fdr")) # adjusted = 508
```

#### 5.1. Volcano plots

A volcano plot is a scatter plot that is used to identify changes between two conditions. It plots the significance **(− log10 P-value)** on the y-axis and the fold-change on the x axes.

###### 38. Generate a volcano plot.

```{r}
volcanoplot(fit3, coef = 1, style = "p-value", highlight = 10, names = fit3$genes$ID, hl.col = "blue",
            xlab = "Log2 Fold Change", ylab = NULL, pch = 16, cex = 0.35)
```

> Points of interest comprehend the highest fold change and statistical significance values. Points where the fold change is log2 = 1 do not present changes between the two conditions. Values on the left represent down regulation while values on the right represent up regulated expression.
